module-services/service-cellular/Pdu.cpp

```diff
@@ -0,0 +1,972 @@
+        }
+
+        PushFields pushFields;
+        if (!Parser::parseUintvar(octets, pushFields.headersLen)) {
+            return std::nullopt;
+        }
+        Octets headersOctets = octets.subOctets(pushFields.headersLen);
+        if (!Parser::parseContentType(headersOctets, pushFields.contentType)) {
+            return std::nullopt;
+        }
+
+        if (!isKnown(pushFields.contentType,
+                     {constants::ApplicationVndWapMmsMessage, "application/vnd.wap.mms-message"})) {
+            return std::nullopt; // unsupported Push type
+        }
+
+        // Push headers
+        {
+            std::array<HeaderValue, 1> pushValues;
+            if (!Parser::parseHeaders(
+                    headersOctets,
+                    std::array<KnownHeader, 1>{{{constants::XWapApplicationId, "X-Wap-Application-ID"}}},
+                    pushValues)) {
+                return std::nullopt;
+            }
+            pushFields.applicationId = std::move(pushValues[0]);
+        }
+
+        if (!isKnown(pushFields.applicationId, {constants::XWapApplicationMmsUa, "x-wap-application:mms.ua"})) {
+            return std::nullopt; // unsupported WAP Push app
+        }
+
+        // MMS headers
+        // https://www.openmobilealliance.org/release/MMS/V1_3-20110913-A/OMA-TS-MMS_ENC-V1_3-20110913-A.pdf, p21
+        MmsCommonFields mmsCommonFields;
+        MmsNotificationFields mmsNotificationFields;
+        {
+            std::array<HeaderValue, 3> mmsValues;
+            if (!Parser::parseHeaders(
+                    octets,
+                    std::array<KnownHeader, 3>{{{constants::XMmsMessageType, "X-Mms-Message-Type"},
+                                                {constants::MmsFrom, "From"},
+                                                {constants::XMmsContentLocation, "X-Mms-Content-Location"}}},
+                    mmsValues)) {
+                return std::nullopt;
+            }
+            mmsCommonFields.messageType           = std::move(mmsValues[0]);
+            mmsNotificationFields.fromAddress     = std::move(mmsValues[1]);
+            mmsNotificationFields.contentLocation = std::move(mmsValues[2]);
+        }
+
+        // https://www.openmobilealliance.org/release/MMS/V1_3-20110913-A/OMA-TS-MMS_ENC-V1_3-20110913-A.pdf, p56
+        if (!isKnown(mmsCommonFields.messageType, {constants::MmsMNotificationInd, "m-notification-ind"})) {
+            return std::nullopt; // unsuported MMS type
+        }
+
+        // https://www.openmobilealliance.org/release/MMS/V1_3-20110913-A/OMA-TS-MMS_ENC-V1_3-20110913-A.pdf, p55
+        Text address;
+        if (std::holds_alternative<Octets>(mmsNotificationFields.fromAddress)) {
+            Octets octets      = std::get<Octets>(mmsNotificationFields.fromAddress);
+            std::uint8_t token = 0;
+            if (!octets.next(token)) {
+                return std::nullopt;
+            }
+            static const std::uint8_t AddressPresentToken = 128;
```

Comment:
why this has to be static? can't it be constexpr?

Reply:
Old habit. I'll change it. ;)